src/background/websocket-manager.ts

```diff
@@ -0,0 +1,361 @@
+      }
+    }
+
+    for (const scriptHash of scriptsHashes) {
+      await this.forceUpdate(scriptHash, network);
+    }
+  }
+}
+
+// utility function converting an output script hex to a scripthash (reversed sha256 of the script)
+// scripthash is used by electrum protocol to identify a script to watch
+export function reverseAndHash(script: string): string {
+  return crypto.sha256(Buffer.from(script, 'hex')).reverse().toString('hex');
+}
+
+// this function is triggered each time a new status is received for a scripthash
+// it fetches the new txs from history and updates the utxos and transactions state
+async function updateScriptHash(
+  ws: ElectrumWS,
+  network: NetworkString,
+  scriptHash: string,
+  store: Store
+): Promise<void> {
+  const historyResp = (await ws.request(
+    'blockchain.scripthash.get_history',
+    scriptHash
+  )) as History;
+  const historyDiff = selectHistoryDiff(historyResp)(store.getState());
+  const accountID = selectAccountIDByScriptHash(scriptHash, network)(store.getState());
+  const blindKeyGetter = getPrivateBlindKeyGetter(store, network);
+  const walletScripts = await getAllWalletScripts(store, network);
+
+  for (const { txID, height } of historyDiff.confirmedTxs) {
+    if (height <= 0) continue;
+    const blockHeaderHex = (await ws.request('blockchain.block.header', height)) as string;
+    const blockHeader = deserializeBlockHeader(blockHeaderHex);
+    const action = confirmTx(txID, blockHeader.timestamp, network);
+    store.dispatch(action);
+  }
+
+  // for each new tx which is not in the state, we fetch the transaction and unblind it
+  for (const { txID, height } of historyDiff.newTxs) {
+    if (selectIsKnownTx(txID)(store.getState())) continue;
+    try {
+      store.dispatch(pushUpdaterLoader());
+      const txFromElectrum = await fetchTx(ws, txID, height > 0 ? height : undefined);
+      const { unblindedTx } = await unblindTransaction(txFromElectrum, blindKeyGetter);
+      const displayTx = toDisplayTransaction(unblindedTx, walletScripts, networks[network]);
+
+      // add the new tx to the store
+      store.dispatch(addTx(accountID, displayTx, network));
+
+      // check if any utxos were spent in this tx
+      for (const input of unblindedTx.vin) {
+        const [utxoInStore, accountID] = selectUtxoByOutpoint(input, network)(store.getState());
+        if (utxoInStore) store.dispatch(deleteUtxo(accountID, input.txid, input.vout, network));
+      }
+
+      const state = store.getState();
+      const allTransactions = selectTransactions(...selectAllAccountsIDs(state))(state);
+
+      // check if we have any new utxos
+      const walletOutputs = unblindedTx.vout
+        .filter((o) => {
+          if (isConfidentialOutput(o)) {
+            return isUnblindedOutput(o);
+          }
+
+          return walletScripts.includes(o.prevout.script.toString('hex'));
+        })
+        .filter(
```

Comment:
we can avodin two filter which are also bit expensive, if we have a data strcuture in storage that simplify the script existance lookup?